tomcat-8.5/src/org/dogtagpki/tomcat/JSSContext.java

```diff
@@ -0,0 +1,102 @@
+import org.mozilla.jss.ssl.javax.JSSParameters;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JSSContext implements org.apache.tomcat.util.net.SSLContext {
+    public static Logger logger = LoggerFactory.getLogger(JSSContext.class);
+
+    private javax.net.ssl.SSLContext ctx;
+    private JSSParameters params;
+
+    public JSSContext(String alias, List<String> negotiableProtocols) {
+        logger.debug("TomcatJSSContext(" + alias + ", ...");
+
+        params = new JSSParameters();
+        params.setAlias(alias);
+        params.setProtocols(negotiableProtocols.toArray(new String[0]));
+    }
+
+    public void init(javax.net.ssl.KeyManager[] kms, javax.net.ssl.TrustManager[] tms, java.security.SecureRandom sr) throws java.security.KeyManagementException {
+        logger.debug("TomcatJSSContext.init(...)");
+        try {
+            ctx = javax.net.ssl.SSLContext.getDefault();
+        } catch (java.security.NoSuchAlgorithmException e) {
+            throw new java.security.KeyManagementException(e.getMessage(), e);
+        }
+        ctx.init(kms, tms, sr);
+    }
+
+    public javax.net.ssl.SSLEngine createSSLEngine() {
+        logger.debug("TomcatJSSContext.createSSLEngine()");
+        JSSEngine eng = (JSSEngine) ctx.createSSLEngine();
+        eng.setSSLParameters(params);
+        return eng;
+    }
+
+    public javax.net.ssl.SSLSessionContext getServerSessionContext() {
+        logger.debug("TomcatJSSContext.getServerSessionContext()");
+        return ctx.getServerSessionContext();
+    }
+
+    public javax.net.ssl.SSLServerSocketFactory getServerSocketFactory() {
+        logger.debug("TomcatJSSContext.getServerSocketFactory()");
+        return ctx.getServerSocketFactory();
+    }
+
+    public javax.net.ssl.SSLParameters getSupportedSSLParameters() {
+        logger.debug("TomcatJSSContext.getSupportedSSLParameters()");
+        return params;
+    }
+
+    public java.security.cert.X509Certificate[] getCertificateChain(java.lang.String alias) {
+        logger.debug("TomcatJSSContext.getCertificateChain(" + alias + ")");
+
+        Exception last = null;
+        KeyManagerFactory kmf = null;
+
+        try {
+            kmf = KeyManagerFactory.getInstance("NssX509");
+        } catch (Exception e) {
+            throw new RuntimeException("Unable to load NSS KeyManagerFactory: " + e.getMessage(), e);
+        }
+
+        for (KeyManager km : kmf.getKeyManagers()) {
+            try {
+                if (km instanceof JSSKeyManager) {
+                    JSSKeyManager jkm = (JSSKeyManager) km;
+                    java.security.cert.X509Certificate[] ret = jkm.getCertificateChain(alias);
+                    if (ret != null) {
+                        return ret;
+                    }
+                }
+            } catch (Exception e) {
+                last = e;
+            }
+        }
+
+        throw new RuntimeException(last.getMessage(), last);
```

Comment:
This might throw NPE if the cert chain is not found and there's no exception during the lookup. I'd say we should rethrow the exception as soon as it's found, then return null or throw a different exception if the cert chain is not found.

Reply:
See previous discussion. This issue goes away if we move the code into JSS.